wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketServerPersonalAccessTokenFetcher.java

```diff
@@ -11,18 +11,88 @@
+import static java.lang.String.valueOf;
+
+import com.google.common.collect.ImmutableSet;
+import java.net.URL;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.factory.server.bitbucket.server.BitbucketPersonalAccessToken;
+import org.eclipse.che.api.factory.server.bitbucket.server.BitbucketServerApi;
+import org.eclipse.che.api.factory.server.bitbucket.server.BitbucketUser;
 import org.eclipse.che.api.factory.server.scm.PersonalAccessToken;
 import org.eclipse.che.api.factory.server.scm.PersonalAccessTokenFetcher;
+import org.eclipse.che.api.factory.server.scm.exception.ScmBadRequestException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmCommunicationException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmItemNotFoundException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmUnauthorizedException;
+import org.eclipse.che.commons.env.EnvironmentContext;
 import org.eclipse.che.commons.subject.Subject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Bitbucket implementation for {@link PersonalAccessTokenFetcher}. Right now returns {@code null}
  * for all possible SCM URL-s (which is valid value) but later will be extended to fully featured
  * class.
  */
 public class BitbucketServerPersonalAccessTokenFetcher implements PersonalAccessTokenFetcher {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(BitbucketServerPersonalAccessTokenFetcher.class);
+
+  private static final String TOKEN_NAME_TEMPLATE = "che-token-<%s>-<%s>";
+  private final BitbucketServerApi bitbucketServerApi;
+  private final URL apiEndpoint;
+
+  @Inject
+  public BitbucketServerPersonalAccessTokenFetcher(
+      BitbucketServerApi bitbucketServerApi, @Named("che.api") URL apiEndpoint) {
+    this.bitbucketServerApi = bitbucketServerApi;
+    this.apiEndpoint = apiEndpoint;
+  }
+
   @Override
-  public PersonalAccessToken fetchPersonalAccessToken(Subject cheUser, String scmServerUrl) {
-    return null;
+  public PersonalAccessToken fetchPersonalAccessToken(Subject cheUser, String scmServerUrl)
+      throws ScmUnauthorizedException, ScmCommunicationException {
+    if (!bitbucketServerApi.isConnected(scmServerUrl)) {
+      LOG.debug("not a  valid url {} for current fetcher ", scmServerUrl);
+      return null;
+    }
+
+    String tokenName = format(TOKEN_NAME_TEMPLATE, cheUser.getUserId(), apiEndpoint.getHost());
+    try {
+      BitbucketUser user = bitbucketServerApi.getUser(EnvironmentContext.getCurrent().getSubject());
+      LOG.debug("Current bitbucket user {} ", user);
+      // cleanup existed
+      List<BitbucketPersonalAccessToken> existedTokens =
+          bitbucketServerApi
+              .getPersonalAccessTokens(user.getSlug())
+              .stream()
+              .filter(p -> p.getName().equals(tokenName))
+              .collect(Collectors.toList());
```

Comment:
why not `forEach(t -> {})` ?

Reply:
deletePersonalAccessTokens throws a lot of checked exceptions. I will not look pretty.